### DISCUSSION TITLE
[help] mirai dispatcher timeout using targets with crew

### DISCUSSION THREAD

**Discussion by alexpiper** [ropensci/targets#1112] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Hi, im experiencing an issue when attempting to run a targets pipeline with both a local crew controller and renv activated, i get a "no targets found" error. The pipeline runs fine when the crew controller is omitted and renv activated, or when the crew controller is active and renv is deactivated. It just does not work together (see reprex below).

**crew + renv (fails)**
```
# _targets.R file
library(targets)  
library(crew)  
tar_option_set(controller = crew_controller_local())
list(
      tar_target(data, data.frame(x = sample.int(1e3))),
      tar_target(summary, mean(data$x, na.rm = TRUE))
)
```
```
> tar_make()
✖ no targets found [7.406 seconds]
Error:
! Error running targets::tar_make()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: connection to local process timed out after 5s
```

**renv with no crew (works)**
```
# _targets.R file
library(targets)
library(crew)
list(
      tar_target(data, data.frame(x = sample.int(1e3))),
      tar_target(summary, mean(data$x, na.rm = TRUE))
)
```
```
> tar_make()
▶ start target data
● built target data [0 seconds]
▶ start target summary
● built target summary [0 seconds]
▶ end pipeline [0.125 seconds]
```
**crew with no renv (works)**
```
# _targets.R file
library(targets)  
library(crew)  
tar_option_set(controller = crew_controller_local())
list(
      tar_target(data, data.frame(x = sample.int(1e3))),
      tar_target(summary, mean(data$x, na.rm = TRUE))
)
```
```
> renv::deactivate()
Restarting R session...
> library(targets)
> library(crew)
> tar_make()
▶ start target data
● built target data [0.016 seconds]
▶ start target summary
● built target summary [0 seconds]
▶ end pipeline [11.937 seconds]
```

**Session info**

```
> sessionInfo()
R version 4.2.0 (2022-04-22 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19044)

Matrix products: default

locale:
[1] LC_COLLATE=English_Australia.utf8  LC_CTYPE=English_Australia.utf8    LC_MONETARY=English_Australia.utf8
[4] LC_NUMERIC=C                       LC_TIME=English_Australia.utf8    

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
[1] crew_0.4.0    targets_1.2.2

loaded via a namespace (and not attached):
 [1] igraph_1.4.1      rstudioapi_0.14   knitr_1.42        magrittr_2.0.3    nanonext_0.9.2   
 [6] tidyselect_1.2.0  R6_2.5.1          rlang_1.1.1       fansi_1.0.4       tools_4.2.0      
[11] data.table_1.14.8 xfun_0.37         sessioninfo_1.2.2 utf8_1.2.3        cli_3.6.1        
[16] yaml_2.3.7        base64url_1.4     digest_0.6.31     tibble_3.2.1      lifecycle_1.0.3  
[21] processx_3.8.2    mirai_0.9.1       callr_3.7.3       vctrs_0.6.3       ps_1.7.5         
[26] codetools_0.2-19  glue_1.6.2        getip_0.1-3       compiler_4.2.0    pillar_1.9.0     
[31] backports_1.4.1   renv_0.17.3-88    pkgconfig_2.0.3 
```

Any help with this would be greatly appreciated.

**Comment by wlandau** [comment]
I tried your example on Windows and `renv` with the versions of package you have, and I could not reproduce the error.

There is another error that reads:

> connection to local process timed out after 5s

which seems related to https://github.com/ropensci/targets/discussions/1101 and https://github.com/wlandau/crew/issues/76. If you upgrade to development `mirai` with the following, it may solve the issue:

```r
install.packages(c("nanonext", "mirai"), repos = "https://shikokuchuo.r-universe.dev")
```

If not, then it could be something `renv`-related or `callr`-related, as was the case with https://github.com/ropensci/targets/discussions/506 and https://github.com/rstudio/renv/issues/773. You could try the pipeline without `callr` using `tar_make(callr_function = NULL)`. In general, it is helpful to see how much you can isolate the issue by peeling back more layers. If you can reproduce the error without `targets`, it will be a lot simpler to understand. If you try to run the example from https://wlandau.github.io/crew/ in your `renv` project and it hits the same error, then you will have narrowed it down.

**Comment by alexpiper** [comment]
Thanks for your help with this Will,

Installing the development versions of mirai and nanonext didnt help, neither did running it without callr, but i think ive narrowed it down to the `daemons` function in `mirai`, where the dispatcher times out when loading larger renv lockfiles (i.e., fails when lockfile contains 260 packages), but runs successfully with a smaller renv lockfile (110 packages). 

Interestingly `daemons` runs successfully when `dispatcher = FALSE`, so it seems to be related to the dispatcher in particular. The time limit for launching the dispatcher seems to be hardcoded as 5000ms in `mirai:::.timelimit`, so perhaps if this timelimit can be edited it may fix the issue. I will open an issue over the mirai github.

See further examples below of the issue occuring with crew and mirai

**Crew**
```
controller <- crew_controller_local(
  name = "example",
  workers = 2,
  seconds_idle = 10,
  
)
controller$start()
Error in launch_and_sync_daemon(sock = sock, urld, parse_dots(...), url,  : 
 sync with dispatcher or daemon failed after 5s
```

**Mirai**
```
# Running daemons with dispatcher (Fails)
daemons(1, dispatcher = TRUE)
Error in launch_and_sync_daemon(sock = sock, urld, dots, n, urlc) : 
  sync with dispatcher or daemon failed after 5s

# Running daemons without dispatcher (works)
daemons(1, dispatcher = FALSE)
[1] 1
```

**Reply by wlandau** [reply]
In that case, you could speed up `renv` initialization by setting the following environment variables in your `.Renviron` file (see `usethis::edit_r_environ()`):

```r
RENV_CONFIG_RSPM_ENABLED=false
RENV_CONFIG_SANDBOX_ENABLED=false
RENV_CONFIG_SYNCHRONIZED_CHECK=false
```

@shikokuchuo, IIRC `mirai` transitioned away from `--vanilla` to allow the versions of `mirai` to agree between the host and dispatcher. But could the timeout error message recommend this workaround, or could the timeout length be configurable with an argument or environment variable?

**Reply by shikokuchuo** [reply]
Technically you only need the daemons to be configured identically to host. Dispatcher could have a minimal configuration as long as it loads. However it didn't before when `--vanilla` was passed as the package library was not found in certain cases.

The vanilla flag implies `--no-site-file, --no-init-file, --no-environ`. Ideally it would be possible to find a set of flags that work, but such a solution may also not be robust as options could be set anywhere.

I have implemented an alternative solution in mirai 0.9.1.9009. The sync timeout can be made configurable for `daemons()` by setting a numeric value for 'dispatcher'. The argument overloading here is not just to reduce clutter, but avoids having arguments which are only applicable if other arguments are set. @wlandau this should work for crew?

**Reply by wlandau** [reply]
> Ideally it would be possible to find a set of flags that work, but such a solution may also not be robust as options could be set anywhere.

Yeah, that's tricky. `renv` relies on the project-level `.Rprofile` file. Suppressing that project `.Rprofile` would avoid `renv`, but as you say, it could also prevent `mirai` from being found.

> I have implemented an alternative solution in mirai 0.9.1.9009. The sync timeout can be made configurable for daemons() by setting a numeric value for 'dispatcher'. The argument overloading here is not just to reduce clutter, but avoids having arguments which are only applicable if other arguments are set. @wlandau this should work for crew?

Thanks @shikokuchuo. I think that would certainly help a lot, and I have plans to expose this new feature in `crew`: https://github.com/wlandau/crew/issues/109. I think it would also help to make the `mirai`-level timeout message more informative so users know where to look.

**Reply by wlandau** [reply]
Or, if timeout errors are classed, I guess I could just catch them in `crew` and write a custom error message myself.

**Reply by shikokuchuo** [reply]
That shouldn't be necessary. They're already better than before, but I can make them more tailored to the message origin.

_All done in mirai 0.9.1.9010_

**Reply by wlandau** [reply]
Thanks.

In `crew`, I decided to temporarily set `RENV_CONFIG_SANDBOX_ENABLED` and `RENV_CONFIG_SYNCHRONIZED_CHECK` to `"false"` just before starting the dispatcher. Hopefully that will help cases like these. I can't think of another use case where an R process would take more than 5 seconds to start.

Alternatively, what if `mirai` went back to `--vanilla` and other flags which could speed up initialization, but also pass on the library path of the parent so the child process could load packages from the same place?

**Reply by shikokuchuo** [reply]
It's a good idea, however dispatcher can also be remote, and then the lib paths may be different.

**Reply by shikokuchuo** [reply]
I guess having to supply a timeout is a bit hit and miss though. I am thinking of reverting the argument to logical only. But have NA as the option to waive the timeout. This won't be completely no timeout otherwise it may block the session if dispatcher really doesn't launch, but let's say times out after 20s.

**Reply by shikokuchuo** [reply]
I've gone ahead and implemented this in shikokuchuo/mirai@23c4cf8 This is foremost a UI issue - if 20s is the default, this is really unacceptably long to wait. There is a helpful error message after the 5s timeout to set the argument value to NA to allow for longer. Then at least if it still times out (due to a connection issue etc.) then people are expecting the longer wait.

**Reply by shikokuchuo** [reply]
> Alternatively, what if `mirai` went back to `--vanilla` and other flags which could speed up initialization, but also pass on the library path of the parent so the child process could load packages from the same place?

@wlandau I've gone and implemented this in shikokuchuo/mirai@2f03d0a. Just to see if it could be done. When launching dispatcher locally from a `daemons()` call, we can reliably locate the libpath for `mirai` and pass it along so dispatcher can run in a `--vanilla` R session for efficiency. Are you able to test that things still work in your cluster environment that failed before?

**Reply by wlandau** [reply]
Thanks for pursuing this, I think it will speed up launch times if successful. 

Currently when I install the latest dev `mirai`/`nanonext` I get an error launching the dispatcher:

```r
Error in `as.environment("package:mirai")`: no item called "package:mirai" on the search list
Backtrace:
 1. x$start()
      at test-crew_clean.R:14:2
 2. self$client$start()
      at crew/R/crew_controller.R:174:8
 4. mirai (local) `<fn>`(...)
 5. mirai:::launch_and_sync_daemon(...)
 6. mirai:::launch_daemon(..., tls = tls)
 7. base::file.path(...)
```

I believe the error is at https://github.com/shikokuchuo/mirai/blob/2f03d0a84693695c3ac9a9af0a66e4818617ac93/R/mirai.R#L1282. `crew` uses the namespace but does not attach `mirai`, so it may not be in the search path.

**Reply by shikokuchuo** [reply]
Ah good catch! I've just updated - do try the new build shikokuchuo/mirai@3529241

**Reply by wlandau** [reply]
Seem to work now, thanks @shikokuchuo!

**Reply by shikokuchuo** [reply]
Fantastic! That should allow you not to have to hack the env vars for renv specifically right?

**Reply by wlandau** [reply]
Yup, works great! I removed those envvar hacks from `crew`, and the dispatcher starts in about 0.6s even when the renv library takes around 8 seconds to load on the host.

**Reply by shikokuchuo** [reply]
FYI in shikokuchuo/mirai@97ebf51 I find the libpath more robustly through the `nanonext` library which is loaded even if `mirai` is not attached. This is not as efficient vs. if `mirai` is already attached but consistent in the case it is not. Trying to find the path of an unattached package is just slow... see the hoops we have to jump through...

**Reply by shikokuchuo** [reply]
And this should be the final iteration: [6e5175b](https://github.com/shikokuchuo/mirai/commit/6e5175bd2394e7eff1d5e0d264c84d726b63c61a). I have essentially re-implemented `find.package()` in a one-liner that takes 3 microseconds to run vs. 80.

**Reply by wlandau** [reply]
Thanks, @shikokuchuo! Still works on my end (and quickly).

**Reply by alexpiper** [reply]
Piping in here to thank you both for the quick response to this issue, the issue is resolved with the latest development version of `mirai`.
